scheduler/feasible.go

```diff
@@ -237,6 +259,83 @@ func (iter *ProposedAllocConstraintIterator) satisfiesDistinctHosts(option *stru
 	return true
 }
 
+// satisfiesBalance checks if the allocation on this node would make the zones
+// unbalanced, this implies a greater than 1 difference between the lowest, and the
+// highest zone
+func (iter *ProposedAllocConstraintIterator) satisfiesBalance(option *structs.Node) bool {
+	// Check if there is no constraint set.
+	if !(iter.jobBalance || iter.tgBalance) {
+		return true
+	}
+
+	if len(iter.job.Datacenters) == 0 {
+		iter.ctx.Logger().Print("[ERR] Job needs at least 1 datacenter to use balance")
+		return false
+	}
+
+	// fill the map with all the dc's in the selected datacenter
+	balanceMap := make(map[string]int)
+
+	for _, dc := range iter.job.Datacenters {
+		balanceMap[dc] = 0
+	}
+
+	// get all the nodes
+	nodeIter, err := iter.ctx.State().Nodes()
+	if err != nil {
+		iter.ctx.Logger().Print("[ERR] Failed to get nodes")
+		return false
+	}
+
+	for {
+		next := nodeIter.Next()
+		if next == nil {
+			break
+		}
+
+		node := next.(*structs.Node)
+
+		// proposed allocations
+		proposed, err := iter.ctx.ProposedAllocs(node.ID)
+		if err != nil {
+			iter.ctx.Logger().Printf("[ERR] scheduler.dynamic-constraint: failed to get proposed allocations: %v", err)
+			return false
+		}
+
+		for _, alloc := range proposed {
+			jobCollision := alloc.JobID == iter.job.ID
+			taskCollision := alloc.TaskGroup == iter.tg.Name
+
+			// skip jobs not in this job or taskgroup (for jobBalance/tgBalance)
+			if !(jobCollision && (iter.jobBalance || taskCollision)) {
+				continue
+			}
+
+			// skip allocation with DesiredStatus other than running
+			if alloc.DesiredStatus != structs.AllocDesiredStatusRun {
+				continue
+			}
+
+			balanceMap[node.Datacenter]++
+		}
+	}
+
+	min := math.MaxInt32
```

Comment:
I think this logic should be detect the desired count of the task group and then divide it by the number of DCs.
For any DC that has less than that max you would return true and for any over you return false.

There is an edge case to consider that if the count is 10 and len(dcs) == 3, the code would have to detect that only one DC should be allowed to get to count 4.

There is a problem with the current implementation because imagine we have `count = 4` and we have 2 nodes in DC1, and one in DC2 and DC3. You would expect this to work but if the input was `[1, 1, 2, 3]`. It would return false on the second node in DC1.